opcua/common/subscription.py

```diff
@@ -201,22 +201,54 @@ def _get_node(self, nodeid):
     def _get_filter_from_event_type(self, eventtype):
         eventtype = self._get_node(eventtype)
         evfilter = ua.EventFilter()
-        for property in get_event_properties_from_type_node(eventtype):
+        evfilter.SelectClauses = self._select_clauses_from_evtype(eventtype)
+        evfilter.WhereClause = self._where_clause_from_evtype(eventtype)
+        return evfilter
+
+    def _select_clauses_from_evtype(self, evtype):
+        clauses = []
+        for property in get_event_properties_from_type_node(evtype):
             op = ua.SimpleAttributeOperand()
-            op.TypeDefinitionId = eventtype.nodeid
+            op.TypeDefinitionId = evtype.nodeid
             op.AttributeId = ua.AttributeIds.Value
             op.BrowsePath = [property.get_browse_name()]
-            evfilter.SelectClauses.append(op)
-        return evfilter
-
-    def subscribe_events(self, sourcenode=ua.ObjectIds.Server, evtype=ua.ObjectIds.BaseEventType):
+            clauses.append(op)
+        return clauses
+
+    def _where_clause_from_evtype(self, evtype):
+        cf = ua.ContentFilter()
+        el = ua.ContentFilterElement()
+        # operands can be ElementOperand, LiteralOperand, AttributeOperand, SimpleAttribute
+        op = ua.SimpleAttributeOperand()
+        op.AttributeId = ua.AttributeIds.NodeId
+        el.FilterOperands.append(op)
+        for subtypeid in [st.nodeid for st in self._get_subtypes(evtype)]:
+            op = ua.LiteralOperand()
+            op.Value = subtypeid
+            el.FilterOperands.append(op)
+        el.FilterOperator = ua.FilterOperator.InList
+
+        cf.Elements.append(el)
+        return cf
+
+    def _get_subtypes(self, parent, nodes=None):
+        if nodes is None:
+            nodes = [parent]
+        for child in parent.get_children(refs=ua.ObjectIds.HasSubtype):
+            nodes.append(child)
+            self._get_subtypes(child, nodes)
+        return nodes
+
+    def subscribe_events(self, sourcenode=ua.ObjectIds.Server, evtype=ua.ObjectIds.BaseEventType, evfilter=None):
         """
         Subscribe to events from a node. Default node is Server node.
         In most servers the server node is the only one you can subscribe to.
+        if evfilter is provided, evtype is ignored
         Return a handle which can be used to unsubscribe
         """
         sourcenode = self._get_node(sourcenode)
-        evfilter = self._get_filter_from_event_type(evtype)
+        if evfilter is None:
```

Comment:
I think 'is None' is correct here since None is the default value. they are equivalen here, but one could alway think that a user could pass a structture with a __boo__ method return False although they wanted really to pass it as arg: http://stackoverflow.com/questions/7152441/python-if-not-val-vs-if-val-is-none